### PR TITLE
DOC: Add description of default bounds to linprog

### LINKED ISSUE #6139: scipy.optimize.linprog failed to find a feasible starting point while solution exists
Following code will cause this problem

```
from scipy import array, vstack, hstack
from scipy.optimize import linprog
from scipy.linalg import solve

f1 = array([[1., 0., 0.], [-1000., 0., - 1000.]])
g1 = array([5.00000000e+00, -1.00000000e+04])
f2 = -array([[0., 1000000., 1010000.]])
g2 = -array([10000000.])

sol = solve(vstack([f1, f2]), hstack([g1, g2-1]))
assert (f1@sol == g1).all()
assert (f2@sol < g2).all()
ret = linprog([1]*3, f2, g2, f1, g1, bounds=[(None, None)]*3)
print(ret)
```

sol is obviously a feasible solution, but the result message is "Optimization failed. Unable to find a feasible starting point."

Is it a bug or I did something wrong?

I'm using scipy 0.17.0 from http://www.lfd.uci.edu/~gohlke/pythonlibs/#scipy under Windows 10/Python3.5.

### PR DESCRIPTION
**Pull request by Kai-Striega**
closes #6139 

By default linear programming bounds variables to be non-negative causing some confusion for users not familiar with LP. This PR explicitly states the default bounds (``lb = 0``, ``ub = np.inf``) and that they may be set by using the ``bounds`` parameter.

@mdhaber @ilayn

### CODE REVIEW

File: scipy/optimize/_linprog.py [hunk 1]
```diff
@@ -165,10 +165,13 @@ def linprog(c, A_ub=None, b_ub=None, A_eq=None, b_eq=None,
...
-        Minimize:     c^T * x
+        Minimize:     c^T @ x
 
-        Subject to:   A_ub * x <= b_ub
-                      A_eq * x == b_eq
+        Subject to:   A_ub @ x <= b_ub
+                      A_eq @ x == b_eq
+                      lb <= x <= ub
+
+        Where ``lb = 0`` and ``ub = np.inf`` unless set in ``bounds``.
```
Comment (ilayn) [inline scipy/optimize/_linprog.py]: This line needs one less level of indentation to get out of code block.

File: scipy/optimize/_linprog_ip.py [hunk 2]
```diff
@@ -539,9 +539,9 @@ def _ip_hsd(A, b, c, c0, alpha0, beta, maxiter, disp, tol,
     r"""
     Solve a linear programming problem in standard form:
 
-    minimize:     c'^T * x'
+    minimize:     c'^T @ x'
 
-    subject to:   A * x' == b
+    subject to:   A @ x' == b
                   0 < x' < oo
```
Comment (ilayn) [inline scipy/optimize/_linprog_ip.py]: Maybe mentioning inf explicitly rather than ASCII art?

File: scipy/optimize/_linprog_ip.py [hunk 3]
```diff
@@ -780,9 +780,9 @@ def _linprog_ip(
 
     Linear programming is intended to solve problems of the following form::
 
-        Minimize:     c^T * x
+        Minimize:     c^T @ x
 
-        Subject to:      A * x == b
+        Subject to:      A @ x == b
                          0 <= x < oo
```
Comment (ilayn) [inline scipy/optimize/_linprog_ip.py]: Same as above about infinity

File: scipy/optimize/_linprog_simplex.py [hunk 4]
```diff
@@ -134,17 +134,13 @@ def _solve_simplex(T, n, basis, maxiter=1000, phase=2, status=0, message='',
...
 
-    subject to
-
-    .. math::
-
-        Ax = b
-        x_i >= 0
-        b_j >= 0
+        subject to:   A @ x' == b
+                      0 < x' < oo
```
Comment (ilayn) [inline scipy/optimize/_linprog_simplex.py]: same infinity

File: scipy/optimize/_linprog_simplex.py [hunk 5]
```diff
@@ -310,12 +306,12 @@ def _linprog_simplex(c, c0, A, b, maxiter=1000, disp=False, callback=None,
...
-    simplex algorithm.::
+    simplex algorithm.
 
-        minimize:     c^T * x
+        minimize:     c'^T @ x'
 
-        subject to:  A * x == b
-                    0 <= x < oo
+        subject to:   A @ x' == b
+                      0 < x' < oo
```
Comment (ilayn) [inline scipy/optimize/_linprog_simplex.py]: One more infinity :)

File: scipy/optimize/_linprog.py (hunk 1 above)
Comment (ilayn) [reply scipy/optimize/_linprog.py]: and may be start without the capitalization?

**Review by ilayn — COMMENTED** [review]
Minor cosmetic comments. 

I noticed that the block in 

https://github.com/scipy/scipy/blob/576c79963dc716ddb37fac572ea616ba64e61644/scipy/optimize/_linprog_simplex.py#L309

is not rendered correctly and (I think) needs another empty line at the bottom. My guess is the immediate next block start is disturbing its rendering.

**Comment by Kai-Striega** [comment]
@ilayn All fixed. I've changed `` 0 < x < oo`` to `` x >= 0`` this aligns the documentation with how it is normally written.

**Comment by mdhaber** [comment]
There are still some issues; you can see what the documentation looks like here:
https://8824-1460385-gh.circle-artifacts.com/0/html-scipyorg/optimize.html

For instance, there is a missing line break in the linprog documentation constraints. There are several differences between the interior point and simplex problem statement definitions, including  remaining ASCII art and formatting as `code`. There are extra ’ characters that I don't think are intended in the simplex documentation.

**Comment by Kai-Striega** [comment]
@mdhaber could you tell me how you got the link? The failing test isn't related to the changes, but I'd like to have a look over the built docs to see if it's formatted properly.

**Comment by ilayn** [comment]
Link is availabile inside Circle CI under the artifacts tab (I think you also need to login). This is the latest one https://8837-1460385-gh.circle-artifacts.com/0/html-scipyorg/optimize.html

**Comment by rgommers** [comment]
> Link is availabile inside Circle CI under the artifacts tab (I think you also need to login).

I think that that's only visible if you have commit rights on scipy Github repo (not 100% sure). The alternative is to build locally; if you already have an in-place build that you've been working up a PR in then it's basically `cd doc && make html` - html docs are then accessible from `doc/build/html/index.html`).

About the rendered version of this PR: that didn't look perfect before, and the extra lines added aren't rendered correctly. I think you want to use ``::` so the problem is rendered as code. Will comment inline.

**Comment by mdhaber** [comment]
@rgommers yes, i think you have to have commit rights. Or at least be logged into CI. I know I couldn't see artifacts before I got commit rights (and looked CI and GitHub account).

There used to be a trick for generating the link yourself based on a base URL and the four digit number in CircleCI, but it didn't work last time I tried.

@kai-striega if you have trouble building docs locally, just push and ping me; I'll send the link.

**Comment by Kai-Striega** [comment]
@mdhaber could you take a look? I've tried building them locally and now seem O.K.

**Comment by mdhaber** [comment]
I'll try to restrict my attention to whether this issue is resolved and clear mistakes. 

On the `linprog` documentation, yes, issue is resolved. Although we also switched to `@` from `*`, which I think makes `^T` unnecessary.

It says upper bound is `np.inf` at the top and `None` elsewhere, but isn't explicit that they're equivalent so it seems inconsistent.

In any case, "By default bounds are (0, None) (non-negative)" needs a period.

In simplex documentation.
```
Minimize: c^T @ x

Subject to: A @ x == b
                0 >= x
```
the words are in teletype. They're weren't in simplex. 
And it's `x >= 0`

In the callback description:
"The phase of the optimization being executed. In phase 1 a basic feasible solution is sought and the T has an additional row representing an alternate objective function."
From the reader's perspective, what is `T`? And it's not even part of the `OptimizeResult` anymore, so probably don't need to mention it at all.

Reference links (like [3]) should be linked by including underscores... I think before? See examples in the docstrings with underscores and how they link.

For status messages, all of them are the short versions except:
`Serious numerical difficulties which could not resolved using
    a more robust, albeit less efficient, solver encountered`
which looks weird because `not` turns green. How about just `Numerical difficulties encountered.`

https://8890-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.optimize.OptimizeResult.html#scipy.optimize.OptimizeResult
renders strangely at the bottom. 
Also some of these are not applicable for linprog. 

But I am getting carried away. Yes, the issue about default bounds not being obvious is addressed, but the things above might be changed while we're at it. Of course there are lots of other things I might change, so they can also wait.

**Comment by Kai-Striega** [comment]
@mdhaber I've changed the docstring for simplex (again) and reformatted the docstring example to match, set the upper bound as ``None`` and set [3] to be hyperrefs. I think the other changes can wait until you are finished with the revised simplex method.

**Comment by mdhaber** [comment]
The documentation says the objective function is `cT @ x`; shouldn't this just be `c @ x` because that is the code used to perform the dot product between `c` and `x`? That was my point before, that `^T` wasn't necessary since `@` suggests that we are writing code, not math notation, and in code we write `c @ x`.

"By default bounds are (0, None) (non-negative) If a sequence containing a single tuple is provided..." needs a period between "(non-negative)" and "If"

I haven't checked whether it's present in the  `scipy.optimize.OptimizeResult`, but I remember that `_postprocess` should output the residuals of the equality constraints (just like it outputs the slack for the inequality constraints). If that is included in the `OptimizeResult` it should be documented in all four files (linprog, interior-point, simplex, and verbose callback).

[Interior point documentation](https://8901-1460385-gh.circle-artifacts.com/0/html-scipyorg/optimize.linprog-interior-point.html) says:
"Minimize a linear objective function subject to linear equality constraints, linear inequality constraints, and simple bounds using the interior point method of [4]."
but that is no longer reflected in the mathematical problem description. 
I should have brought this up before, but I'm conflicted in which should be changed. On one hand., the mathematical problem description in [Interior point documentation](https://8901-1460385-gh.circle-artifacts.com/0/html-scipyorg/optimize.linprog-interior-point.html) is accurate for `_linprog_ip`. However, the reader might be confused that the problem description for the main `linprog` routine does not match this mathematical problem description, only the text description. 

`[scipy.optimize.linprog_verbose_callback](https://8901-1460385-gh.circle-artifacts.com/0/html-scipyorg/generated/scipy.optimize.linprog_verbose_callback.html#scipy.optimize.linprog_verbose_callback)` still shows:
`Serious numerical difficulties which could not resolved using a more robust, albeit less efficient, solver encountered`

**Comment by Kai-Striega** [comment]
@mdhaber I've made changes to you mentioned above, plus some in the remaining function docs (hopefully for the last time)

I've also changed the documentation of the ``interior-point`` method, to match the mathematical formulation (i.e. s.t. equality and non-negativity constraints) then modified the notes section in linprog saying the problem is converted to standard form.

**Comment by mdhaber** [comment]
@Kai-Striega That approach sounds OK, but:
"Before applying either method a presolve procedure based on [8]_ and converted attempts to standard form." 
doesn't make sense. Note that 
"After presolve, the problem is transformed to standard form by converting the (tightened) simple bounds to upper bound constraints, introducing non-negative slack variables for inequality constraints, and expressing unbounded variables as the difference between two non-negative variables."
already appeared at the end of the notes. I think that is a more logical place to put that bit about standard form.
But neither are in a very prominent place; still I think that only users familiar with linear programming standard form will understand why the problem form changes. Perhaps there could be a note in each of the methods documentations calling attention to the difference and explaining that it is the result of the problem provided to `linprog` being converted to standard form.

Also `con` is still missing from the `OptimizeResult` returned by `linprog`.

"Solve the following linear programming problem via a two-phase simplex algorithm of the form:"
doesn't strictly make sense.

Sorry to point out more things : / I know you want this "simple" thing to be done : )

**Comment by mdhaber** [comment]
I'll take a look this weekend and hopefully merge!

**Pull request merged**